lib/pg_online_schema_change/orchestrate.rb

```diff
@@ -156,32 +167,39 @@ def disable_vacuum!
             autovacuum_enabled = false, toast.autovacuum_enabled = false
           );
         SQL
-        Query.run(client.connection, sql)
+        Query.run(client.connection, sql, true)
       end
 
       def run_alter_statement!
+        # re-uses transaction with serializable
         statement = Query.alter_statement_for(client, shadow_table)
         logger.info("Running alter statement on shadow table",
                     { shadow_table: shadow_table, parent_table: client.table })
-        Query.run(client.connection, statement)
+        Query.run(client.connection, statement, true)
 
         Store.set(:dropped_columns_list, Query.dropped_columns(client))
         Store.set(:renamed_columns_list, Query.renamed_columns(client))
       end
 
-      # Begin the process to copy data into copy table
-      # depending on the size of the table, this can be a time
-      # taking operation.
       def copy_data!
-        logger.info("Copying contents..", { shadow_table: shadow_table, parent_table: client.table })
+        # re-uses transaction with serializable
+        # Begin the process to copy data into copy table
+        # depending on the size of the table, this can be a time
+        # taking operation.
+        logger.info("Clearing contents of audit table before copy..",
+                    { shadow_table: shadow_table, parent_table: client.table })
+        Query.run(client.connection, "TRUNCATE #{audit_table}", true)
```

Comment:
I think we need to use `DELETE FROM` here; otherwise, `TRUNCATE` takes out an `AccessExclusive` lock on the shadow table for the duration of the transaction which will include the `INSERT INTO`. That will have the knock on effect of blocking writes on the original table.

Reply:
Well, interestingly, I wasn't able to reproduce my suspected issue with a running pgbench.

Oh, i see why. There's an existing transaction running already when you try to start one in serializable mode:

```
2022-02-16 21:26:04.592 CST [46534] LOG:  duration: 0.018 ms  statement: COMMIT;
2022-02-16 21:26:04.593 CST [46534] LOG:  duration: 0.003 ms  statement: BEGIN;
2022-02-16 21:26:04.593 CST [46534] WARNING:  there is already a transaction in progress
2022-02-16 21:26:04.593 CST [46534] LOG:  duration: 0.016 ms  statement: BEGIN ISOLATION LEVEL SERIALIZABLE
```

Reply:
If it's difficult to remove the nested transactions, we may be able to use `SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;` https://www.postgresql.org/docs/9.6/sql-set-transaction.html

Reply:
Yeah the truncate is on the audit table, since nothing is using it at the moment, it should be fine. But I removed it just in case, since the code then becomes easy to reason about too. 

Good idea on the `SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;`, totally missed that.

We do suffer from multiple `BEGIN` few places, so that should get fixed in future releases after a refactor. 

Updated with changes.

Reply:
No problem acquiring the lock for the `TRUNCATE`, it's just that the lock is held for the duration of the transaction and not the `TRUNCATE` statement, so it would block writes to the original table because the trigger wouldn't be able to write into the shadow table with that `AcessExclusive` lock being held while the `INSERT INTO` is happening.

Reply:
Oh! Yes absolutely. You are correct. Nice catchy

Reply:
Should be good now.